src/joybus.c

```diff
@@ -57,293 +71,224 @@ static volatile struct SI_regs_s * const SI_regs = (struct SI_regs_s *)0xa480000
 static void * const PIF_RAM = (void *)0x1fc007c0;
 
 /**
- * @brief Wait until the SI is finished with a DMA request
+ * @brief A message to be sent to JoyBus, with its completion callback. 
+ */
+typedef struct {
+    uint64_t input[JOYBUS_BLOCK_DWORDS] __attribute__((aligned(16)));  ///< input message
+    void (*callback)(uint64_t* output);                                ///< callback for completion
+} joybus_msg_t;
+
+#define MAX_JOYBUS_MSGS            8    ///< Maximum number of pending joybus messages
+#define JOYBUS_STATE_IDLE          0    ///< Joybus state: idle (no pending messages)
+#define JOYBUS_STATE_SENDING       1    ///< JoyBus state: sending a message to PIF
+#define JOYBUS_STATE_RECEIVING     2    ///< JoyBus state: receiving a reply from PIF
+
+/** @brief Joybus temporary output buffer */
+static uint64_t joybus_outbuf[JOYBUS_BLOCK_DWORDS] __attribute__((aligned(16)));
+/** @brief Joybus current state (either #JOYBUS_STATE_IDLE, #JOYBUS_STATE_SENDING or #JOYBUS_STATE_RECEIVING) */
+static volatile int joybus_state;
+/** @brief Joybus pending messages (ring buffer) */
+static joybus_msg_t joybus_msgs[MAX_JOYBUS_MSGS];
+/** @brief Pending messages write index */
+static volatile int msgs_widx;
+/** @brief Pending messages read index */
+static volatile int msgs_ridx;
+
+static void si_interrupt(void);
+
+/**
+ * @brief Initialize the joybus subsystem
  */
-static void __SI_DMA_wait( void )
+__attribute__((constructor))
```

Comment:
I think this is OK but if someone wants to disable this maybe we should provide a way?

Reply:
Which case are you thinking of? The constructor is not called if any other function of joybus.c is not called. So if you're not using joybus at all, the constructor is not called. Otherwise, you need the interrupt to be able to use joybus.

The alternatives are:

1) initialize the interrupt lazily on the first joybus_exec_async call. That requires a check every call. I would say the constructor approach achieves the same without the very little overhead, but I agree it's more obscure.
2) add an explicit joybus_init function, and call it from all joybus-related init functions (eg: controller_init). This works as well but breaks usages of joybus_exec in the wild as no init function was required before. joybus_exec is marked as experimental API though.

Any preference?

Reply:
> The constructor is not called if any other function of joybus.c is not called.
I totally forgot about that. That makes this much less important and I think the constructor is fine 👍